stats: add histograms for request/response headers and body sizes

ranjithkumar007 opened the pull request:
Signed-off-by: Ranjith Kumar <ranjith.dakshana2015@gmail.com>

Commit Message:  stats: add histograms for request/response headers and body sizes
Additional Description:  Created a new struct for optional cluster stats. Moved timeout budget stats and added request response headers and body stats in the new struct.
Risk Level: Low
Testing: Added test cases
Docs Changes: added
Release Notes: added
Fixes #10308 , Fixes #3621

ranjithkumar007 commented:
PTAL @nezdolik

nezdolik commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -772,6 +772,8 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(RequestHeaderMapPtr&& he
                                connection_manager_.read_callbacks_->connection().dispatcher());
   request_headers_ = std::move(headers);
 
+  clusterInfo()->stats().upstream_rq_headers_size_.recordValue(request_headers_->byteSize());
```
you can probably access cluster_info object via `connection_manager_.read_callbacks_->clusterInfo()` and not add extra method to ActiveStream api.

ranjithkumar007 replied:
Got `clusterinfo()` by `connection_manager_.read_callbacks_->upstreamHost()->cluster()`. Couldn't find `clusterInfo()` directly part of `read_callbacks_`.

nezdolik commented on docs/root/configuration/upstream/cluster_manager/cluster_stats.rst:
```diff
@@ -77,6 +77,7 @@ Every cluster has a statistics tree rooted at *cluster.<name>.* with the followi
   upstream_rq_retry_limit_exceeded, Counter, Total requests not retried due to exceeding :ref:`the configured number of maximum retries <config_http_filters_router_x-envoy-max-retries>`
   upstream_rq_retry_success, Counter, Total request retry successes
   upstream_rq_retry_overflow, Counter, Total requests not retried due to circuit breaking or exceeding the :ref:`retry budget <envoy_v3_api_field_config.cluster.v3.CircuitBreakers.Thresholds.retry_budget>`
+  upstream_rq_headers_size, Histogram, Request headers size in bytes per upstream
```
i would put this stat in alphabetical order within their corresponding subgroup (eg relateive to `upstream_rq_`)

nezdolik commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -772,6 +772,8 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(RequestHeaderMapPtr&& he
                                connection_manager_.read_callbacks_->connection().dispatcher());
   request_headers_ = std::move(headers);
 
+  clusterInfo()->stats().upstream_rq_headers_size_.recordValue(request_headers_->byteSize());
```
we need to verify in tests that value is recorded once per entire filter chain

ranjithkumar007 commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -6049,6 +6050,35 @@ TEST_F(HttpConnectionManagerImplTest, NewConnection) {
+  std::shared_ptr<Upstream::MockClusterInfo> cluster_info{
+      new NiceMock<Upstream::MockClusterInfo>()};
+
+  EXPECT_CALL(*decoder_filters_[0]->callbacks_, clusterInfo())
```
![Screenshot from 2020-06-19 23-16-20](https://user-images.githubusercontent.com/17480353/85197376-38a50e80-b2fe-11ea-86b2-c6abc15f2106.png)
Getting an error here. How to add clusterinfo to callbacks?

ranjithkumar007 replied:
I think `decoder_filters_[0]` is a  `MockStreamDecoderFilter` which should ideally have set clusterinfo [callback](https://github.com/envoyproxy/envoy/blob/74290ef76a76fbbf50f072dc33438791f93f68c7/test/mocks/http/mocks.cc#L43).

ranjithkumar007 commented:
All checks have passed. PTAL @nezdolik

jmarantz commented on test/integration/stats_integration_test.cc:
```diff
@@ -286,8 +288,8 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithFakeSymbolTable) {
-  EXPECT_MEMORY_EQ(m_per_cluster, 44491);
-  EXPECT_MEMORY_LE(m_per_cluster, 44811);
+  EXPECT_MEMORY_EQ(m_per_cluster, 47707);
+  EXPECT_MEMORY_LE(m_per_cluster, 47771);
```
can you round this up  47771 to 48000? IDK how these exact values got derived but really these _LE values are by nature approximate as they measure byte-count consumption across platforms and compile-modes.

We're just having the _LE versions to help catch major changes during local tests, without waiting for the CI "release" build where the _EQ version is tested.

ranjithkumar007 replied:
Thanks, will update the values.

jmarantz commented on test/integration/stats_integration_test.cc:
```diff
@@ -273,6 +273,8 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithFakeSymbolTable) {
   // 2020/05/13  10531    44425       44600   Refactor resource manager
   // 2020/05/20  11223    44491       44600   Add primary clusters tracking to cluster manager.
   // 2020/06/10  11561    44491       44811   Make upstreams pluggable
+  // 2020/06/23  11559    47707       47771   stats: add histogram for request headers size per
```
This is the reason that these memory tests exist. We can now have the conversation about whether the value of having a per-cluster histogram for request sizes is worth consuming 3216 bytes per cluster. Consider that deployments may have 10s of thousands of clusters.

It might have that value but it's not obvious.

ranjithkumar007 replied:
Great to have such tests tracking memory consumption.

ranjithkumar007 replied:
Should we add a new entry like `enable_upstream_rq_stats` in [cluster config](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#config-cluster-v3-cluster) ?

ranjithkumar007 replied:
Made the stat optional.

jmarantz commented on test/integration/stats_integration_test.cc:
```diff
@@ -349,8 +353,8 @@ TEST_P(ClusterMemoryTestRunner, MemoryLargeClusterSizeWithRealSymbolTable) {
-  EXPECT_MEMORY_EQ(m_per_cluster, 36603);
-  EXPECT_MEMORY_LE(m_per_cluster, 36923);
+  EXPECT_MEMORY_EQ(m_per_cluster, 39739);
+  EXPECT_MEMORY_LE(m_per_cluster, 39803);
```
round this to 40000

nezdolik commented:
After brief discussion with @jmarantz, proposed histogram should be enabled by config flag, as it is expensive to track histogram per cluster with large number of clusters. @ranjithkumar007 here is example of histogram controlled by config flag: https://github.com/envoyproxy/envoy/pull/9227

ranjithkumar007 commented:
> here is example of histogram controlled by config flag: #9227

Thanks @nezdolik . So, we need a new entry in [cluster config](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/cluster.proto#config-cluster-v3-cluster). 
I was wondering how to add such entry that would fit well for #11559 as well.
Should we have seperate entries like `track_upstream_rq_headers_size`, `track_upstream_rq_body_size` ? or a dict like `track_upstream_rq_stats : {'headers_size': true, 'body_size' : false}`?

nezdolik commented:
Think the first approach is preferred.

ranjithkumar007 commented:
Now that this stat is optional, there is no need to update memory limits in `stats_integration_test.cc`?

jmarantz commented:
> Does stats_integration_test.cc include optional stats as well while tracking memory consumption?

No, so you can back your change to stats_integration_test.cc, which should hopefully fix CI. Thanks!

jmarantz commented:
/wait

ranjithkumar007 commented:
> you can back your change to stats_integration_test.cc

The memory consumption values([1](https://dev.azure.com/cncf/envoy/_build/results?buildId=43349&view=logs&j=8d802004-fbbb-5f17-b73e-f23de0c1dec8&t=86512d3a-caa5-5e3f-9bce-4070faa3b830&l=6475), [2](https://dev.azure.com/cncf/envoy/_build/results?buildId=43349&view=logs&j=8d802004-fbbb-5f17-b73e-f23de0c1dec8&t=86512d3a-caa5-5e3f-9bce-4070faa3b830&l=6561)) from CI are exactly 256 bytes higher than the values on `master`([1](https://github.com/envoyproxy/envoy/blob/master/test/integration/stats_integration_test.cc#L289), [2](https://github.com/envoyproxy/envoy/blob/master/test/integration/stats_integration_test.cc#L352)). 

`absl::optional<T>`  occupied memory(so 256 bytes overhead) even if it was empty.

ranjithkumar007 commented:
/wait

ranjithkumar007 commented on docs/root/configuration/upstream/cluster_manager/cluster_stats.rst:
```diff
@@ -314,3 +314,17 @@ Statistics for monitoring effective host weights when using the
+Request Response Size statistics
+--------------------------------
+
+If :ref:`request response size statistics are tracked <envoy_v3_api_field_config.cluster.v3.Cluster.track_request_response_sizes>` is, 
```
TODO: Remove `is`.

ranjithkumar007 commented:
windows ci failure unrelated to PR

jmarantz commented on source/common/upstream/upstream_impl.h:
```diff
@@ -634,6 +638,7 @@ class ClusterInfoImpl : public ClusterInfo, protected Logger::Loggable<Logger::I
   Stats::ScopePtr stats_scope_;
   mutable ClusterStats stats_;
   Stats::IsolatedStoreImpl load_report_stats_store_;
+  const absl::optional<ClusterRequestResponseSizeStats> request_response_size_stats_;
```
256 extra bytes for an absl::optional seems egregious. Is it reserving the entire space for the object, rather than a pointer? 

My suggestion is to (a) use a unique_ptr rather than absl::optional here -- you can still return the optional in the accessor above, and (b) combine this and ClusterTimeoutBudgetStats below into a single; possibly nested structure.

That way the default case will take minimal overhead of just one pointer for all aggregated optional stats.

ranjithkumar007 replied:
> Is it reserving the entire space for the object, rather than a pointer?

Qouting from an [article](https://abseil.io/tips/123), `empty absl::optional uses as much memory as a populated one`. It says that that memory usage of `absl::optional<Bar>` is always `sizeof(Bar) + sizeof(bool)` irrespective of whether it is null or not. However, with `std::unique_ptr<Bar>`, memory usage is ` sizeof(Bar*) when null` and `sizeof(Bar*) + sizeof(Bar) otherwise`

ranjithkumar007 replied:
> you can still return the optional in the accessor above

Is this to make code changes minimal?

ranjithkumar007 replied:
> one pointer for all aggregated optional stats

```
struct OptionalStats{
    std::unique_ptr<ClusterRequestResponseSizeStats> p1;
    std::unique_ptr<ClusterTimeoutBudgetStats> p2;
    ...
};

std::unique_ptr<OptionalStats> p3;
```
Something like above? which would occupy 8 bytes for `unique_ptr<OptionalStats >` in general case.

I guess this would also mean we need to have a nested definition for the config of Optional stats as well, something like `track_optional_stats: {timeout_budget: true, request_response_sizes: true}`.

jmarantz replied:
RE "still return the optional" -- it's not to minimize code changes; that makes sense at the interface. But it should be an optional reference, not a reference to the optional object.

So, for. your sample code, `std::unique_ptr<OptionalStats> p3;` looks perfect.

I think inside your OptionalStats struct you could use optionals or unique_ptr. I'm OK either way. Main disadvantage of unique_ptr inside is the double-indirection, but it's probably fine.

nezdolik replied:
nit: maybe name it `OptionalClusterStats`

ranjithkumar007 replied:
Thank you..I think using unique_ptr inside strict would be better as the use cases of each optional stat are quite different. And if we use optional inside struct then it would add up memory for all stats even if we need to track a subset of stats.

ranjithkumar007 commented on docs/root/configuration/upstream/cluster_manager/cluster_stats.rst:
```diff
@@ -314,3 +314,17 @@ Statistics for monitoring effective host weights when using the
   min_entries_per_host, Gauge, Minimum number of entries for a single host
   max_entries_per_host, Gauge, Maximum number of entries for a single host
+
+.. _config_cluster_manager_cluster_stats_request_response_sizes:
```
test

nezdolik commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -836,6 +836,11 @@ message Cluster {
   core.v3.TypedExtensionConfig upstream_config = 48;
+
+  // If track_request_response_sizes is true, then the :ref:`histograms
+  // <config_cluster_manager_cluster_stats_request_response_sizes>`  tracking header and body sizes
```
let's align docs with what is implemented

ranjithkumar007 replied:
Can we merge #11766 with the current pr? Else aligning docs would need several redundant changes on both prs.

ranjithkumar007 commented:
/wait

ranjithkumar007 commented on test/common/upstream/upstream_impl_test.cc:
```diff
@@ -2235,6 +2235,34 @@ TEST_F(ClusterInfoImplTest, OneofExtensionProtocolOptionsForUnknownFilter) {
+    connect_timeout: 0.25s
+    type: STRICT_DNS
+    lb_policy: ROUND_ROBIN
+    track_request_response_sizes: true
```
Ah.. missed this.. will update config here

ranjithkumar007 commented on docs/root/configuration/upstream/cluster_manager/cluster_stats.rst:
```diff
@@ -314,3 +314,17 @@ Statistics for monitoring effective host weights when using the
+
+.. _config_cluster_manager_cluster_stats_request_response_sizes:
+
+Request Response Size statistics
```
TODO: Update docs once other parts are finalized.

ranjithkumar007 commented on docs/root/version_history/current.rst:
```diff
@@ -137,6 +137,7 @@ New Features
 * server: add the option :option:`--drain-strategy` to enable different drain strategies for DrainManager::drainClose().
 * server: added :ref:`server.envoy_bug_failures <server_statistics>` statistic to count ENVOY_BUG failures.
 * stats: added the option to :ref:`report counters as deltas <envoy_v3_api_field_config.metrics.v3.MetricsServiceConfig.report_counters_as_deltas>` to the metrics service stats sink.
+* stats: added histogram to :ref:`cluster stats <config_cluster_manager_cluster_stats_request_response_sizes>` that tracks request headers size.
```
TODO: Update

ranjithkumar007 commented:
/wait

ranjithkumar007 commented:
@repokitteh bot :rotating_light:, don't remove waiting label :neutral_face:

ranjithkumar007 commented:
/wait

ranjithkumar007 commented:
Ci tests passeD except Marcos which I guess is flaky and unrelated to Pr.

@jmarantz , weirdly stats_integration_test have passed without any modifications.

jmarantz commented:
Can you look at the CI for docs? Failure looks real (I don't know much about it myself)

ranjithkumar007 commented:
> Can you look at the CI for docs? Failure looks real (I don't know much about it myself)
> 

Yeah, I will update docs.

ranjithkumar007 commented:
/wait

ranjithkumar007 commented:
/wait

nezdolik commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -911,3 +907,17 @@ message UpstreamConnectionOptions {
+  // If request_response_sizes is true, then the :ref:`histograms
+  // <config_cluster_manager_cluster_stats_request_response_sizes>`  tracking header and body sizes
+  // of requests and responses will be published.
+  bool request_response_sizes = 2;
```
we may need more granular flags here, would this flag enable tracking of 4 histograms at same time? Some users may be for example only interested in tracking requests related stats.

ranjithkumar007 replied:
Currently, It enables all histograms under `struct ClusterRequestResponseSizeStats`. Yes, we might need more granularity.

nezdolik replied:
I think we should at least split it on req/resp path.

ranjithkumar007 replied:
I am Ok with separating the stats into `RequestSizeStats` and `ResponseSizeStats`. 
@jmarantz views on this?

jmarantz replied:
I'm OK with that but I don't see a strong need for that much fine-grained control. My reasoning is that by default I think Envoy should run lean, with minimal overhead/cluster. I mildly favor keeping the control coarser for simplicity, but I'm OK either way.

If you want to zoom in on some aspect on a per-cluster basis, you probably want to collect a bunch of data and you may not be sure what you want to to look at, but you are willing to spend memory to do it. You probably won't leave this enabled forever anyway, but you might as learn as much as you can.

For similar reasons I don't typically debug programs compiling only *some* of the files with "-g" (I used to do that when I was in a much more memory-constrained environment).

ranjithkumar007 replied:
Thanks for the detailed thought. 
With that in mind, now I also think it makes sense to leave the granularity as it is.

nezdolik commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -911,3 +907,17 @@ message UpstreamConnectionOptions {
   core.v3.TcpKeepalive tcp_keepalive = 1;
 }
+
+message TrackOptionalClusterStats {
```
imho `Optional` can be dropped from name, as tracking of stats is turned on/off with boolean flags, it implies that it can be specified or not.

ranjithkumar007 replied:
will update.

ranjithkumar007 replied:
Need to improve my nomenclature decisions :100:

nezdolik commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -772,6 +772,14 @@ void ConnectionManagerImpl::ActiveStream::decodeHeaders(RequestHeaderMapPtr&& he
+
+  if (host_ != nullptr) {
+    if (host_->cluster().requestResponseSizeStats().has_value()) {
+      host_->cluster().requestResponseSizeStats()->get().upstream_rq_headers_size_.recordValue(
```
as `requestResponseSizeStats()` may return nullopt, this may be source of npe

nezdolik replied:
nvm, missed if case above

ranjithkumar007 replied:
> source of npe

Fixed the npe earlier. All checks are passing now :rocket:

nezdolik commented on source/common/router/router.cc:
```diff
@@ -949,7 +949,7 @@ void Filter::onUpstreamTimeoutAbort(StreamInfo::ResponseFlag response_flags,
         dispatcher.timeSource().monotonicTime() - downstream_request_complete_time_);
 
-    cluster_->timeoutBudgetStats()->upstream_rq_timeout_budget_percent_used_.recordValue(
+    cluster_->timeoutBudgetStats()->get().upstream_rq_timeout_budget_percent_used_.recordValue(
```
do we need to check here if `timeoutBudgetStats() != nullopt` ?

nezdolik commented on test/common/upstream/upstream_impl_test.cc:
```diff
@@ -2235,6 +2235,45 @@ TEST_F(ClusterInfoImplTest, OneofExtensionProtocolOptionsForUnknownFilter) {
                             "extension_protocol_options can be specified");
 }
+
+TEST_F(ClusterInfoImplTest, TestTrackRequestResponseSizes) {
```
rename `TestTrackRequestResponseSizes` to highlight purpose of the test, eg `TestTrackRequestResponseSizesNotSetInConfig`

ranjithkumar007 replied:
Currently `TestTrackRequestResponseSizes` has tests both for not set and set in config. Should i seperate them into 2 functions?

nezdolik replied:
could you provide tests for not set in config, set to true in config and set to false in config?

ranjithkumar007 replied:
ok.. I think i missed setting to `false` case. I will add it.

nezdolik commented on test/common/upstream/upstream_impl_test.cc:
```diff
@@ -2235,6 +2235,45 @@ TEST_F(ClusterInfoImplTest, OneofExtensionProtocolOptionsForUnknownFilter) {
+
+  auto cluster = makeCluster(yaml_disabled);
+  // By default, histograms tracking request/response sizes are not published.
+  EXPECT_FALSE(cluster->info()->requestResponseSizeStats().has_value());
```
do we also need to check that tracking of timeout budgets is not enabled?

ranjithkumar007 replied:
Nice Catch. Thanks.

ranjithkumar007 replied:
on second thought, we don't need to check that tracking of timeout budgets is not enabled here as it is already checked in `TestTrackTimeoutBudgets`.

nezdolik commented on include/envoy/upstream/upstream.h:
```diff
@@ -650,13 +656,28 @@ struct ClusterCircuitBreakersStats {
+};
+
+using ClusterRequestResponseSizePtr = std::unique_ptr<ClusterRequestResponseSizeStats>;
+using ClusterRequestResponseSizeOptRef =
```
is there is a reason we need to keep ptr and optRef types? would not be ptr enough?

ranjithkumar007 replied:
Based on discussion [here](https://github.com/envoyproxy/envoy/pull/11559#discussion_r446968854).
Also, If we return a unique_ptr<T>, then there is the issue with transfer of ownership which I don't think is intended.
If we return a pointer T* directly, the caller has to know if it is responsible for freeing the memory or someone else is. That's not something we are concerned with when it's a reference.

nezdolik replied:
got it

nezdolik commented on source/common/router/upstream_request.cc:
```diff
@@ -92,7 +92,8 @@ UpstreamRequest::~UpstreamRequest() {
         std::chrono::duration_cast<std::chrono::milliseconds>(end_time - start_time_);
     parent_.cluster()
         ->timeoutBudgetStats()
```
here don't see quarding if `timeoutBudgetStats() != nullopt`

ranjithkumar007 replied:
It is handled using `record_timeout_budget_` in [Line 88](https://github.com/envoyproxy/envoy/blob/e6e61ae66773b603452f13230fd742f7e0ef5251/source/common/router/upstream_request.cc#L88) which is only set if  `timeoutBudgetStats() != nullopt` [here](https://github.com/envoyproxy/envoy/blob/e6e61ae66773b603452f13230fd742f7e0ef5251/source/common/router/upstream_request.cc#L53).

nezdolik commented:
@ranjithkumar007 could you please update links to related issues in PR description? (as now we are addressing 2 issues in same PR)

ranjithkumar007 commented on include/envoy/upstream/upstream.h:
```diff
@@ -851,9 +872,18 @@ class ClusterInfo {
+   * headers/body sizes of request/response for this cluster.
    */
-  virtual const absl::optional<ClusterTimeoutBudgetStats>& timeoutBudgetStats() const PURE;
+
```
Remove extra lines

ranjithkumar007 commented:
/wait

ranjithkumar007 commented:
/wait

ranjithkumar007 commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -6050,6 +6051,226 @@ TEST_F(HttpConnectionManagerImplTest, NewConnection) {
+
+  // invoke encodeHeaders
+  decoder_filters_[0]->callbacks_->encodeHeaders(
+      ResponseHeaderMapPtr{new TestResponseHeaderMapImpl{{":status", "200"}}}, true);
```
response headers are internally mutated and for example in this test case, it ends as `':status', '200' 'date', 'Mon, 06 Jul 2020 06:08:55 GMT' 'server', ''`. So, final headers size is `49`. Is this okay or should we return original response headers size(`10`)?

nezdolik replied:
you could verify size with final headers and put explanatory comment about internal headers being appended

ranjithkumar007 replied:
Thanks.. Will update

ranjithkumar007 commented:
/wait

ranjithkumar007 commented:
/wait

ranjithkumar007 commented:
Windows ci build failure looks unrelated to PR.

jmarantz commented:
Please do not force-push in envoy github PRs -- it makes reviews difficult because comments get lost.

Just merge master.

ranjithkumar007 commented:
> Please do not force-push in envoy github PRs -- it makes reviews difficult because comments get lost.
> 
> Just merge master.

I force pushed on top of a single commit that I made after the review.
I will make sure to merge master every time from now on.  Apologies for this time.

jmarantz commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -610,6 +610,18 @@ ConnectionManagerImpl::ActiveStream::ActiveStream(ConnectionManagerImpl& connect
+  if (host_ != nullptr) {
+    if (host_->cluster().requestResponseSizeStats().has_value()) {
+      // record request body size
+      host_->cluster().requestResponseSizeStats()->get().upstream_rq_body_size_.recordValue(
```
suggest making a temp for readability (and slight perf benefit):
```
  ClusterRequestResponseSizeOptRef req_resp_stats = host_->cluster().requestResponseSizeStats();
  if (req_resp_stats.has_value()) {
    req_resp_stats->get().upstream_rq_body_size_.recordValue(stream_info_.bytesReceived());
...
```
similar patterns below. Actually does it work to use
```
 req_resp_stats->upstream_rq_body_size_.recordValue(stream_info_.bytesReceived());
```
omitting the `->get().` ?

ranjithkumar007 replied:
Will update.
IIRC, omitting the `->get()` doesn't work, `get()` gives access to the stored reference. Will check again.

jmarantz commented on source/common/upstream/upstream_impl.h:
```diff
@@ -577,10 +578,25 @@ class ClusterInfoImpl : public ClusterInfo, protected Logger::Loggable<Logger::I
   Stats::Scope& statsScope() const override { return *stats_scope_; }
+
+  ClusterRequestResponseSizeOptRef requestResponseSizeStats() const override {
+    if (not optional_cluster_stats_ or not(optional_cluster_stats_->request_response_size_stats_)) {
```
although legal C++ we don't typically use this in Envoy style; we would use `!`. Also for the second term, that's a unique_ptr and not a boolean, so it'd be more stylistically consistent to use:
```
optional_cluster_stats_->request_response_size_stats_ != nullptr
```

jmarantz commented on source/common/upstream/upstream_impl.h:
```diff
@@ -577,10 +578,25 @@ class ClusterInfoImpl : public ClusterInfo, protected Logger::Loggable<Logger::I
-    return timeout_budget_stats_;
+
+  ClusterTimeoutBudgetOptRef timeoutBudgetStats() const override {
+    if (not optional_cluster_stats_ or not(optional_cluster_stats_->timeout_budget_stats_)) {
```
ditto

jmarantz commented on test/common/upstream/upstream_impl_test.cc:
```diff
@@ -2332,20 +2389,50 @@ TEST_F(ClusterInfoImplTest, TestTrackTimeoutBudgets) {
   EXPECT_TRUE(cluster->info()->timeoutBudgetStats().has_value());
+  EXPECT_EQ(
+      Stats::Histogram::Unit::Unspecified,
+      cluster->info()->timeoutBudgetStats()->get().upstream_rq_timeout_budget_percent_used_.unit());
```
consider a temp for `cluster->info()->timeoutBudgetStats()->get().upstream_rq_timeout_budget_percent_used_` or similar.

jmarantz commented on test/common/upstream/upstream_impl_test.cc:
```diff
@@ -2235,6 +2235,63 @@ TEST_F(ClusterInfoImplTest, OneofExtensionProtocolOptionsForUnknownFilter) {
+
+  auto cluster = makeCluster(yaml);
+  // The stats should be created.
+  EXPECT_TRUE(cluster->info()->requestResponseSizeStats().has_value());
```
ASSERT_TRUE rather than EXPECT_TRUE to avoid crash below.  Then maybe a temp for `cluster->info()->requestResponseSizeStats()->get()`

ggreenway commented on api/envoy/config/cluster/v3/cluster.proto:
```diff
@@ -826,12 +826,8 @@ message Cluster {
-  // request. These show what percentage of a request's per try and global timeout was used. A value
-  // of 0 would indicate that none of the timeout was used or that the timeout was infinite. A value
-  // of 100 would indicate that the request took the entirety of the timeout given to it.
-  bool track_timeout_budgets = 47;
```
You cannot change an existing proto field. The existing field must remain, and a new field added, and optionally deprecate the existing field, but it must be supported according to the deprecation policy.

ranjithkumar007 replied:
not exactly sure, but I gave a try.
I am new to the community. please let me if the changes align the expectations.

ranjithkumar007 replied:
I am new to the community. Not exactly sure if the changes align the expectations, but I gave a try.

ranjithkumar007 commented:
`@ranjithkumar007 dismissed jmarantz’s stale review via f88549a 9 minutes ago`

I am not sure why/when did this happen :confused: ? I just clicked on resolve conversation.

ranjithkumar007 commented:
Will fix ci.
/wait

ranjithkumar007 commented:
/wait

ranjithkumar007 commented:
CI tests are passing except coverage which looks like a flaky issue. 

I read [Breaking change policy](https://github.com/envoyproxy/envoy/blob/master//CONTRIBUTING.md#breaking-change-policy), [API versioning](https://github.com/envoyproxy/envoy/blob/master/api/API_VERSIONING.md) and tried to modify code following those. In the code, I deprecated `track_timeout_budgets` in `v3`, removed it and reserved its field number in `v4alpha`. The generated config for `v4alpha` still has `track_timeout_budgets` in the form of `hidden_envoy_deprecated_track_timeout_budgets`.  Currently, I updated the code to track timeout stats based on `config.has_track_cluster_stats() or config.track_timeout_budgets()` where `config` is `v3 Cluster config` . So, when transitioning from `v3` to `v4`, is this updated to `config.has_track_cluster_stats() or config.hidden_envoy_deprecated_track_timeout_budgets()` and from `v4` to `v5`, do we get rid of `track_timeout_budgets` completely?

jmarantz commented on source/common/upstream/upstream_impl.h:
```diff
@@ -577,10 +578,27 @@ class ClusterInfoImpl : public ClusterInfo, protected Logger::Loggable<Logger::I
   Stats::Scope& statsScope() const override { return *stats_scope_; }
+
+  ClusterRequestResponseSizeStatsOptRef requestResponseSizeStats() const override {
+    if (optional_cluster_stats_ == nullptr or
```
s/or/||/

ranjithkumar007 replied:
Will update once api review is done.

jmarantz commented on source/common/upstream/upstream_impl.cc:
```diff
@@ -680,10 +686,9 @@ ClusterInfoImpl::ClusterInfoImpl(
-                                ? absl::make_optional<ClusterTimeoutBudgetStats>(
-                                      generateTimeoutBudgetStats(*stats_scope_))
-                                : absl::nullopt),
+      optional_cluster_stats_((config.has_track_cluster_stats() or config.track_timeout_budgets())
```
s/or/||/

jmarantz commented on source/common/upstream/upstream_impl.cc:
```diff
@@ -1088,6 +1093,17 @@ void ClusterImplBase::validateEndpointsForZoneAwareRouting(
+ClusterInfoImpl::OptionalClusterStats::OptionalClusterStats(
+    const envoy::config::cluster::v3::Cluster& config, Stats::Scope& stats_scope)
+    : timeout_budget_stats_(
+          (config.track_cluster_stats().timeout_budgets() or config.track_timeout_budgets())
```
s/or/||/

nezdolik replied:
so if at least one of flags is set to true, histograms will be enabled. We may need to document this expectation or add validation that if `config.track_cluster_stats().timeout_budgets() != config.track_timeout_budgets()` then fail on startup.

ranjithkumar007 replied:
Makes sense. Not sure which one is preferred while deprecating such fields.

nezdolik replied:
after consulting with @htuch, is acceptable to make new field override the old one. The way is implemented now fits this strategy (newer one take priority).

nezdolik commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -6050,6 +6051,230 @@ TEST_F(HttpConnectionManagerImplTest, NewConnection) {
+  EXPECT_CALL(*decoder_filters_[0], decodeComplete());
+
+  filter_callbacks_.upstreamHost(nullptr);
+  // Histograms should not record value when upstream host is not configured.
```
in this test there are no actual checks/stated expectations that histogram is not being created/updated with value.

ranjithkumar007 replied:
If there is record call, then `deliverhistogram..` will be called. Here we don't expect that. So it should mean that there was no record call for any histogram(or else Gmock would throw error for uninteresting call).

I am not sure how to check creation of histograms.

ranjithkumar007 replied:
this test doesn't make sense. Removed.

nezdolik commented on docs/root/configuration/upstream/cluster_manager/cluster_stats.rst:
```diff
@@ -176,7 +176,7 @@ Circuit breakers statistics will be rooted at *cluster.<name>.circuit_breakers.<
 
-If :ref:`timeout budget statistic tracking <envoy_v3_api_field_config.cluster.v3.Cluster.track_timeout_budgets>` is
+If :ref:`timeout budget statistic tracking <envoy_v3_api_field_config.cluster.v3.Cluster.track_cluster_stats>` is
 turned on, statistics will be added to *cluster.<name>* and contain the following:
```
should we keep documentation for deprecated field as well during first deprecation cycle? (cc @htuch )

nezdolik replied:
confirmed with @htuch, documentation for deprecated field should still be in place @ranjithkumar007

ranjithkumar007 replied:
Will update

nezdolik commented:
@envoyproxy/api-shepherds kind ask for api review

snowp commented on source/common/http/conn_manager_impl.cc:
```diff
@@ -611,6 +611,19 @@ ConnectionManagerImpl::ActiveStream::ActiveStream(ConnectionManagerImpl& connect
+    Upstream::ClusterRequestResponseSizeStatsOptRef req_resp_stats =
+        upstream_host->cluster().requestResponseSizeStats();
+    if (req_resp_stats.has_value()) {
+      // record request body size
```
nit: proper grammar in comments (capitalized first letter, period)

here and below

alternatively just get rid of these comments, the code speaks for itself IMO

ranjithkumar007 replied:
Yeah, will update it.

snowp commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -6050,6 +6051,230 @@ TEST_F(HttpConnectionManagerImplTest, NewConnection) {
+  EXPECT_CALL(response_encoder_, encodeHeaders(_, true));
+  expectOnDestroy();
+
+  // invoke encodeHeaders
```
remove, its obvious from the code

snowp commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -6050,6 +6051,230 @@ TEST_F(HttpConnectionManagerImplTest, NewConnection) {
+
+  EXPECT_CALL(response_encoder_, encodeHeaders(_, false));
+
+  // invoke encodeHeaders
```
remove

snowp commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -6050,6 +6051,230 @@ TEST_F(HttpConnectionManagerImplTest, NewConnection) {
+  EXPECT_CALL(response_encoder_, encodeData(_, true));
+  expectOnDestroy();
+
+  // invoke encodeData
```
remove

snowp commented on test/common/http/conn_manager_impl_test.cc:
```diff
@@ -6050,6 +6051,230 @@ TEST_F(HttpConnectionManagerImplTest, NewConnection) {
+  EXPECT_CALL(response_encoder_, encodeData(_, true));
+  expectOnDestroy();
+
+  // invoke encodeData
```
remove

ranjithkumar007 commented:
@jmarantz There is a decline of 256 bytes in memory usage per cluster. On master it is [45003](https://github.com/envoyproxy/envoy/blob/master/test/integration/stats_integration_test.cc#L304). On this PR it shows [44747](https://dev.azure.com/cncf/envoy/_build/results?buildId=45400&view=logs&j=8d802004-fbbb-5f17-b73e-f23de0c1dec8&t=86512d3a-caa5-5e3f-9bce-4070faa3b830&l=6852).

jmarantz commented:
Awesome; one last regold away from passing release tests. Since you changed an existing large structure from absl::optional to unique_ptr, the memory reduction makes sense to me.

Nice job! Fix CI and this LGTM.

htuch commented:
/lgtm api

jmarantz commented:
/azp run

ranjithkumar007 commented:
[hds_integration_test](https://dev.azure.com/cncf/envoy/_build/results?buildId=45502&view=logs&j=5eb754ee-449a-545e-2b62-07fe525d7cbc&t=1110550d-5f5f-55a9-69fd-d02adf65e59a&l=5121) is timing out again. Possibly a flaky test.

jmarantz commented:
/azp run

RE hds_integration_test timing out...not clear whether this is just death by 3 seconds/test-method * lots of test methods, or something is getting stuck. Hopefully we'll catch a faster machine on this run.

nezdolik commented:
PR lgtm apart from the point of adding back docs for deprecated field.

nezdolik commented:
retest

ranjithkumar007 commented:
> Looks like hds flaky test is the reason of failing CI, @ranjithkumar007 could you try to retrigger CI with `/retest` comment?

Yeah, I will do it once I update the docs.

ranjithkumar007 commented:
Thanks everyone!  Good to have first contribution to envoy merged 😀 .
